diffkemp/simpll/passes/FieldAccessFunctionGenerator.cpp

```diff
@@ -0,0 +1,191 @@
+        ValidUserSet.insert(Inst);
+    bool allUsersValid = true;
+    for (Instruction *Inst : Stack) {
+        if (Inst == Stack.back())
+            continue;
+        for (User *U : Inst->users()) {
+            if (ValidUserSet.find(U) == ValidUserSet.end()) {
+                allUsersValid = false;
+                break;
+            }
+        }
+    }
+    if (!allUsersValid) {
+        Stack.clear();
+        return;
+    }
+
+    // Check if all operands in the code are values included in the set created
+    // in the previous step. If not, it would be technically possible to do this
+    // transformation, but this is not supported by this implementation.
+    bool allOperandsValid = true;
+    for (Instruction *Inst : Stack) {
+        if (Inst == Stack.front()) {
+            // The GEP is guaranteed to have all indices constant and the first
+            // value is extracted.
+            continue;
+        }
+        for (Value *Op : Inst->operands()) {
+            if (isa<ConstantData>(Op))
+                // Constants are fine, they can be moved without problem.
+                continue;
+            if (ValidUserSet.find(Op) != ValidUserSet.end())
+                // The value was found in the stack set.
+                continue;
+            allOperandsValid = false;
+            break;
+        }
+    }
+    if (!allOperandsValid) {
+        Stack.clear();
+        return;
+    }
+
+    // Create the function definition.
+    // The generated abstraction receives the source variable as its argument
+    // and returns the result of the field access (both are pointers, as with
+    // simple GEPs).
+    FunctionType *FT = FunctionType::get(Stack.back()->getType(), {
+        Stack.front()->getOperand(0)->getType()
+    }, false);
+    Function *Abstraction = Function::Create(FT,
+            GlobalValue::LinkageTypes::ExternalLinkage,
+            SimpllFieldAccessFunName, &Mod);
+    Abstraction->setMetadata(SimpllFieldAccessMetadata,
+                             Stack.front()->getDebugLoc().get());
+    auto Arg = &*Abstraction->arg_begin();
+
+    // Copy the instructions to the function body.
+    BasicBlock *BB = BasicBlock::Create(Abstraction->getContext(),
+                                        "", Abstraction);
+    for (Instruction *Inst : Stack) {
+        // Move into the new function.
+        if (Inst == Stack.back()) {
+            // The last instruction needs special processing, because it will
+            // be replaced with the function call in the original function.
+            Instruction *Clone = Inst->clone();
+            Clone->setDebugLoc(DebugLoc());
+            auto CallInst = CallInst::Create(FT, Abstraction,
+                    {Stack.front()->getOperand(0)});
+            CallInst->insertAfter(Inst);
+            Inst->replaceAllUsesWith(CallInst);
+            Inst->removeFromParent();
+            BB->getInstList().push_back(Clone);
+            Stack[Stack.size() - 1] = Clone;
```

Comment:
Instead of modifying the last instruction in `Stack`, modify the last instruction in `BB`.